### PR TITLE
Update strict least-upper-bound and subtype to match Lean

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
## Description of changes

Update strict least-upper-bound and subtype to match Lean

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

File: cedar-policy/CHANGELOG.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Deprecated error `TypeErrorKind::ImpossiblePolicy` in favor of warning
   `ValidationWarningKind::ImpossiblePolicy` so future improvements to Cedar
   typing precision will not result in breaking changes. (resolving #539)
+- Changed policy validation to reject comparisons and conditionals between
+  record type that differ in if an attribute is required or optional.
```
Comment (khieta) [inline cedar-policy/CHANGELOG.md:31]: Nit. I just think it sounds better 🤷‍♀️ 
```suggestion
  record type that differ in whether an attribute is required or optional.
```

**Review by khieta — APPROVED** [review]
Looks fine to me. Since this is a breaking change (policies that previously validated no longer validate), make sure to tag this PR with `breaking-change`.

**Review by aaronjeline — APPROVED** [review]

File: cedar-policy/CHANGELOG.md [hunk 2]
```diff
@@ -33,6 +33,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   typing precision will not result in breaking changes. (resolving #539)
 - Made `is_authorized` and `validate` functions in the frontend public, as well as their related structs: `AuthorizationAnswer`, `AuthorizationCall`, `ValidationCall`, `ValidationSettings`, `ValidationEnabled`, `ValidationError`, `ValidationWarning`, `ValidationAnswer`. (#737)
 - Improved `Display` implementation for Cedar schemas, both JSON and human syntax. (#780)
+- Changed policy validation to reject comparisons and conditionals between
+  record type that differ in whether an attribute is required or optional.
```
Comment (khieta) [inline cedar-policy/CHANGELOG.md:37]:
```suggestion
  record types that differ in whether an attribute is required or optional.
```

**Comment by khieta** [comment]
Looking at this again -- what's the error message printed for your test case? I'm worried that the difference between an optional vs. required attribute is not easy to spot, so the error message may be confusing.

**Comment by john-h-kastner-aws** [comment]
> what's the error message printed for your test case?

Something like 

```
unable to find upper bound for types: [{a?: Long,},{a: Long,}]
```

Which isn't great, but we have issue #346 open to improve this sort of error message. I'm not particular concerned about spotting the `?`.

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**